Add support for plex's timeline reporter

chicco-carone opened the pull request:
This pull request includes several changes focusing on improving the player queue handling and enhancing the Plex provider functionality. The most significant changes include updating the player queue state management, adding a timeline reporter to the Plex provider, and refining the handling of stream details and marking tracks as played.

Improvements to player queue handling:
* [`music_assistant/controllers/player_queues.py`]: Updated the `pause` method to set the queue state to PAUSED and signal an update. Added logging and state synchronization in the `_update_queue_from_player` method.

Enhancements to Plex provider:
* [`music_assistant/providers/plex/__init__.py`]: Added initialization of a timeline reporter in the `PlexProvider` class and included a method to close it properly.
* [`music_assistant/providers/plex/__init__.py`]: Refined the `get_stream_details` method to include additional metadata in the returned data dictionary.
* [`music_assistant/providers/plex/__init__.py`]: Improved the `on_streamed` method to handle and log cases where the track data is invalid or missing key information.

New Timeline Reporter Implementation (`music_assistant/helpers/timeline.py`):
* Created a new `PlexTimelineReporter` class that manages playback state synchronization between Music Assistant and Plex servers
* Implements real-time timeline updates for play, pause, and stop states
* Handles track completion and transitions properly with multiple confirmation updates
* Maintains separate client identities for each player with proper device information
* Includes comprehensive error handling and logging
* Uses an efficient async implementation with aiohttp for API communications
* Properly manages resources with cleanup on shutdown
* Provides accurate tracking of playback progress with periodic updates
* Handles edge cases like player power off and network connectivity issues

chicco-carone committed:
Add support for plex's timeline reporter

marcelveldt commented on music_assistant/controllers/player_queues.py:
```diff
@@ -1663,6 +1665,22 @@ def _update_queue_from_player(
         queue.available = player.available
         queue.items = len(self._queue_items[queue_id])
 
+        self.logger.debug(
+            "Player %s state update: player state=%s, queue state=%s",
+            player.display_name,
+            player.state,
+            queue.state,
+        )
+
+        if player.state == PlayerState.PAUSED and queue.state != PlayerState.PAUSED:
+            self.logger.info(
+                "Setting queue state to PAUSED to match player state for %s", player.display_name
+            )
+            queue.state = PlayerState.PAUSED
+            self.signal_update(queue_id)
+            return
+
+        # Normal queue state update from player state
```
Please do not create mixed PR's.
So please create (or propose) a separate PR for adjusting core functionality and a dedicated PR for adjusting a provider.

marcelveldt replied:
Also, I don't think the changes in the queue controller are needed this way. I'll have a look because this smells like a one-line fix in the existing update handler.

marcelveldt commented on music_assistant/helpers/timeline.py:
This helper is unique to Plex, it doesnt belong here at the core helpers at all.
Also I'd like to know exactly why this is all needed because normally we are not conflating players and MA queue states with music sources. So please start with a discussion on discord in one of the dev channels.

Next to that, in general the Plex provider is so broken that I think first the basics need to be fixed and somebody step up as a maintainer before we even consider adding more maintenance burden.

chicco-carone replied:
> I want sure where to put it but moving it to the correct place is a minor issue. About the Plex integration being unmaintained I could be interested in maintaining it since for me Plex is the only source of music and I'm familiar with the library since I used it for other projects. In my experience since I've been using MA only for a few days I didn't notice any issues but if there is a list of things to fix I could into doing that before finishing this pr.

OzGav replied:
Here are the open issues https://github.com/music-assistant/support/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22Plex%22%20

OzGav replied:
Discussions are here https://github.com/orgs/music-assistant/discussions?discussions_q=is%3Aopen+in%3Atitle+plex+-label%3Aimplemented One of them is basically an issue.

marcelveldt commented on music_assistant/providers/plex/__init__.py:
```diff
@@ -944,9 +963,19 @@ async def on_streamed(
         """Handle callback when an item completed streaming."""
 
         def mark_played() -> None:
```
I think what you are interested in is the scrobbling event to report what is playing. Have a look at the lastfm scrobbler for an example (or the template plugin provider).

marcelveldt reviewed (commented):
see the inline comments

chicco-carone commented:
I just moved the timeline file as another provider and set it as a plugin similar to the spotify one or the lastfm one. also I added a fix to the code that parses the singular tracks to remove the hardcoded limit for 99999 and added pagination which should fix the issue mentioned in issue [2711](https://github.com/music-assistant/support/issues/2711) with larger playlist since instead of parsing the whole playlist in one go it splits in pages to avoid timeouts. Now i'm unsure if its better to update this pr with these new changes or make a new one

marcelveldt commented:
It's always best to create smaller PR's that limit to one bugfix or one feature.

OzGav commented:
@chicco-carone Do you still plan to progress this?

chicco-carone commented:
Yea I'll look into it again now that the other fixes for Plex are upstream. I've been currently using this version without the new fixes) in my environment so I have some things that I know I need to fix after using it a bit 
.

chicco-carone commented:
I'm closing this pr to open a new one with the updated main branch and the fixes implemented in a better way. New one on #2228
